### PR TITLE
Enable extra information in passed data

### PR DESCRIPTION
**Pull request by anicke**

### CODE REVIEW

**Comment by ChaitanyaSharma** [comment]
I merged this pull request and tried using the data. However, the data does not render for me. Could you help me figure out what I could be doing wrong?

demo.js file attached

[demo.js.zip](https://github.com/marmelab/EventDrops/files/248055/demo.js.zip)

File: README.md [hunk 1]
```diff
@@ -53,6 +53,25 @@ var data = [
 ];
 ```
 
+More advanced data objects can also be used but this requires us to specify the which attribute that contains the date we want to use. This object is available during the eventColor and eventClick callbacks for example.
```
Comment (jpetitcolas) [inline README.md]: Not sure, but `to specify which attribute is the date` sounds better to me (but I'm not native English (-:)

Comment (jpetitcolas) [inline README.md]: Surround event names with backticks (`) for better formatting

File: README.md [hunk 2]
```diff
@@ -53,6 +53,25 @@ var data = [
...
+
+```js
+var data = [
```
Comment (jpetitcolas) [inline README.md]: Can you split it on several lines, for better readability?

File: README.md [hunk 3]
```diff
@@ -53,6 +53,25 @@ var data = [
...
+  { name: "http requests", dates: [{date: new Date('2014/09/15 13:24:54'), foo: "bar1"}, {date: new Date('2014/09/15 13:25:03'), foo: "bar2"}, {date: new Date('2014/09/15 13:25:05'), foo: "bar1"}, ...] },
+  { name: "SQL queries", dates: [{date: new Date('2014/09/15 13:24:57'), foo: "bar4"}, {date: new Date('2014/09/15 13:25:04'), foo: "bar6"}, {date: new Date('2014/09/15 13:25:04'), foo: "bar2"}, ...] }
+];
+```
+
+Then specify which attribute that contains the date to use.
```
Comment (jpetitcolas) [inline README.md]: Same remark as above

File: README.md [hunk 4]
```diff
@@ -83,6 +102,7 @@ Configurable values:
   - `minScale`: The minimum scaling (zoom out), default to 0.
   - `maxScale`: The maximum scaling (zoom in), default to Infinity.
   - `zoomable`: *true* by default. Enable zoom-in/zoom-out and dragging handlers.
+  - `dateAttribute`: *null* by default. Only required to specify which object field that contains the date to use if the data array contains objects instead of an array of only dates.
```
Comment (jpetitcolas) [inline README.md]: If passing objects, specify which field contains the event date.

File: lib/drawer/drops.js [hunk 5]
```diff
@@ -37,4 +38,11 @@ export default (svg, scales, configuration) => function dropsSelector(data) {
     });
 
     dropLines.exit().remove();
+
+    function getDate(d) {
```
Comment (jpetitcolas) [inline lib/drawer/drops.js]: Can you use a `const` here, and declare it before using it?

File: test/karma/eventDrops.js [hunk 6]
```diff
@@ -79,5 +79,20 @@ describe('d3.chart.eventDrops', () => {
...
         test(true, true);
-    });
+    });
+
+    it('should be possible to supply objects as data', () => {
+        const div = document.createElement('div');
+        const data = [
+            { name: 'foo', dates: [{date: new Date('2010-01-01'), foo: "bar1"}]},
+            { name: 'bar', dates: [] },
+            { name: 'quz', dates: [{date: new Date('2011-01-04'), foo: "bar2"},
```
Comment (jpetitcolas) [inline test/karma/eventDrops.js]: A more readable indentation would be:

``` js
{ name: 'quz', dates: [
    { date: new Date(''), foo: 'bar1' },
    { date: new Date(''), foo: 'bar2 },
]}
```

**Comment by jpetitcolas** [comment]
To better match D3 philosophy, perhaps should we introduce a `date` function returning the date from the object. So, instead of using `dateAttribute`, we should instead use:

``` js
const chart = d3.chart.eventDrops()
    .date(d => d.date);
```

**Comment by jpetitcolas** [comment]
I tested it quickly, and it works well. Thanks a lot! :)

**Comment by anicke** [comment]
I agree that it would be better to use a date function! I'll update the PR.

**Comment by ChaitanyaSharma** [comment]
Nick, I'm still getting weird errors like: "Uncaught TypeError: d3.chart.eventDrops(...).date is not a function"

Would it be possible for you to share a demo JS file as an example that uses additional values in data?

**Comment by anicke** [comment]
@ChaitanyaSharma 
Start the dev server using "make run". Then change the "demo.js" file:
```javascript
for (var j = 0; j < max; j++) {
    var time = (Math.random() * (endTime - startTime)) + startTime;
    event.dates.push({date: new Date(time), dataField: "pass"});
}
...
// create chart function
var eventDropsChart = d3.chart.eventDrops()
    .eventLineColor(function (datum, index) {
        return color(index);
    })
    .date(function(d){
        return d.date;
    })
    .start(new Date(startTime))
    .end(new Date(endTime));
```

**Comment by jpetitcolas** [comment]
Hi @anicke. Any news on this expected PR? :)

Moreover, you need to rebase.

**Comment by anicke** [comment]
PR rebased! Is there anything else I need to do?

File: test/karma/eventDrops.js [hunk 7]
```diff
@@ -80,4 +80,18 @@ describe('d3.chart.eventDrops', () => {
...
+
+    it('should be possible to supply objects as data', () => {
+        const div = document.createElement('div');
+        const data = [
+            { name: 'foo', dates: [{date: new Date('2010-01-01'), foo: 'bar1'}]},
+            { name: 'bar', dates: [] },
+            { name: 'quz', dates: [{date: new Date('2011-01-04'), foo: 'bar2'},
+                                {date: new Date('2012-08-09'), foo: 'bar3'}]}];
+
+        const chart = d3.chart.eventDrops().date(function getDate(d) {return d.date;}).start(new Date('2010-01-25'));
```
Comment (jpetitcolas) [inline test/karma/eventDrops.js]: Use arrow function here for more readability: `.date(d => d.date)`

File: test/karma/eventDrops.js [hunk 8]
```diff
@@ -80,4 +80,18 @@ describe('d3.chart.eventDrops', () => {
         test(false, false);
         test(true, true);
     });
+
+    it('should be possible to supply objects as data', () => {
+        const div = document.createElement('div');
+        const data = [
+            { name: 'foo', dates: [{date: new Date('2010-01-01'), foo: 'bar1'}]},
```
Comment (jpetitcolas) [inline test/karma/eventDrops.js]: Perhaps should we rename the `dates` to `data`, as it is not an array of dates anymore.

File: lib/drawer/drops.js [hunk 9]
```diff
@@ -1,6 +1,14 @@
+
 export default (svg, scales, configuration) => function dropsSelector(data) {
     const dropLines = svg.selectAll('.drop-line').data(data);
 
+    const getDate = function getDateValue(d) {
```
Comment (jpetitcolas) [inline lib/drawer/drops.js]: Or perhaps can we default the config `date` to: `d => d`?

File: README.md [hunk 10]
```diff
@@ -83,6 +108,7 @@ Configurable values:
   - `minScale`: The minimum scaling (zoom out), default to 0.
   - `maxScale`: The maximum scaling (zoom in), default to Infinity.
   - `zoomable`: *true* by default. Enable zoom-in/zoom-out and dragging handlers.
+  - `date`: function that returns the date from each data point when passing objects.
```
Comment (jpetitcolas) [inline README.md]: Add the default value (probably `d => d`).

File: test/karma/eventDrops.js (hunk 8 above)
Comment (jpetitcolas) [inline test/karma/eventDrops.js]: Do not forget to update demo code and documentation. And, bonus, if you can add this backward compatibility break in the changelog... :)

**Comment by jpetitcolas** [comment]
I did an extra review. But it seems quite good! :)

Do not hesitate to ping us when your changes are done: we're not notified on a "simple" push.

**Comment by anicke** [comment]
@jpetitcolas Done, please have a look. This turned out to be bigger change than expected. :)

**Comment by jpetitcolas** [comment]
Yes, that looks good to me. A huge thank for this very requested feature. :)

**Pull request merged**